nixos/modules/services/web-apps/porn-vault.nix

```diff
@@ -0,0 +1,269 @@
+{
+  config,
+  pkgs,
+  lib,
+  ...
+}:
+
+let
+  cfg = config.services.porn-vault;
+  configFormat = pkgs.formats.json { };
+  inherit (lib)
+    mkIf
+    mkEnableOption
+    mkPackageOption
+    mkOption
+    getExe
+    types
+    ;
+in
+{
+  options = {
+    services.porn-vault = {
+      enable = lib.mkEnableOption "Porn-Vault";
+
+      package = lib.mkPackageOption pkgs "porn-vault" { };
+
+      autoStart = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = ''
+          Whether to start porn-vault automatically.
+        '';
+      };
+
+      port = lib.mkOption {
+        type = lib.types.port;
+        default = 3000;
+        description = ''
+          Which port Porn-Vault will use.
+        '';
+      };
+
+      openFirewall = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = ''
+          Whether to open the default ports in the firewall for porn-vault.
+        '';
+      };
+
+      config = {
+        enable = mkEnableOption "Porn-Vault configuration";
+        settings = mkOption {
+          type = configFormat.type;
+          description = ''
+            Configuration for Porn-Vault. The attributes are serialized to JSON in config.json.
+
+            See https://gitlab.com/porn-vault/porn-vault/-/blob/dev/config.example.json
+          '';
+          default = {
+            auth = {
+              password = null;
+            };
+            binaries = {
+              ffmpeg = "ffmpeg";
+              ffprobe = "ffprobe";
+              izzyPort = 8000;
+              imagemagick = {
+                convertPath = "convert";
+                montagePath = "montage";
+                identifyPath = "identify";
+              };
+            };
+            import = {
```

Comment:
As mentioned, this config is quite large and contains many assumptions like directory paths. I guess that the upstream example tries to be exhaustive, i.e. act as documentation for most or all of the config parameters. IMO this should contain only the properties essential for PV to start and have basic functionality successfuly. Things like source paths, metadata matching rules and HW acceleration options should be left to the individual user.

Reply:
Perhaps that could be turned into an example then, so users know all available options, and then a more minimal config could be set as default?

I’d need to do some trial and error in order to see how minimal the config can be without PV failing to run tho, never tried that

Reply:
Just tried and Porn-Vault can’t really work without all keys being there. I tried making a smaller default configuration, but then PV crashes as soon as I open the web app.

So either we keep that as an example but with no default settings at all (which would mean the user would have to provide that themselves, as PV won’t launch without its config file), or we keep the default as it is right now and if that doesn’t work for the user they can change it.

Reply:
> I mean, it's JSON, so we can move it to a separate file and then `lib.importJSON`

That would probably make it impossible for people to declare their pv config on their nixos config tho, right..?

Reply:
I think they are referring to something like `default = lib.importJSON default_config.json`, which shouldn't be any different than including these same values as a Nix literal. Given the type of the option, users should still be able to override parts piecemeal.

Reply:
I see. But wouldn’t that be (even) more convoluted than the current way? 🤔

Reply:
It would certainly make the Nix file itself much smaller so that reviewers and future readers can focus more on the important parts :p

Reply:
If I were to keep it separated, in what folder should that be kept? Modules seem to just be all thrown in the same folder (as opposed to packages which each has its own folder), so where should that config.json go?

Reply:
> If I were to keep it separated, in what folder should that be kept? Modules seem to just be all thrown in the same folder (as opposed to packages which each has its own folder), so where should that config.json go?

Any ideas?

Reply:
Done! Pls take a look at the latest force-push